### PR TITLE
WIP: Add drag and drop example

### PR DESCRIPTION
**Pull request by Susurrus**
This requires the `drag_targets` branch of `gtk` and `gdk` to be used to work at all. It's also not quite correct, but I'm not certain why yet, and these examples are a good test case to make sure that the work in `gtk` and `gdk` is correct. One shows an example of dragging a widget within an app onto another widget. The other will display metadata of anything dropped onto it.

### CODE REVIEW

File: src/text_drop.rs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    let accepted = vec![
+        gtk::TargetEntry::new("text/uri-list", gtk::TargetFlags::empty(), 0),
+//        gtk::TargetEntry::new("text/plain", gtk::TargetFlags::empty(), 1),
+//        gtk::TargetEntry::new("STRING", gtk::TargetFlags::empty(), 2),
+    ];
+    text_view.drag_dest_set(gtk::DEST_DEFAULT_ALL,
+                            &accepted,
+                            gdk::ACTION_COPY | gdk::ACTION_MOVE); // Maybe this should be ACTION_ASK?
+    text_view.connect_drag_data_received(|_, d, _, _, s, tt, t| {
+        println!("-{}", tt);
```
Comment (GuillaumeGomez) [inline src/text_drop.rs]: We're generally not big fans of printing things in stdout when using a graphic application. Use a small window or any equivalent which allows to remain in the application to see the info please.

File: src/drag_and_drop.rs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        println!("Source Drag Data Delete");
+        println!("Delete the data!");
+    });
+
+    // Configure label as drag destination
+    label.drag_dest_set(gtk::DEST_DEFAULT_ALL,
+                        &[],
+                        gdk::ACTION_COPY); // Maybe this should be ACTION_ASK?
+    label.connect_drag_motion(|w, d, _, _, _| {
+        println!("motion");
```
Comment (GuillaumeGomez) [inline src/drag_and_drop.rs]: Same.

File: src/drag_and_drop.rs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    window.add(&vbox);
+    window.show_all();
+
+    // Configure button as drag source
+    button.drag_source_set(gdk::BUTTON1_MASK | gdk::BUTTON2_MASK,
+                           &[],
+                           gdk::ACTION_COPY | gdk::ACTION_MOVE);
+    button.connect_drag_data_get(|_, _, s, _, _| {
+        println!("Source Drag Data Get");
```
Comment (GuillaumeGomez) [inline src/drag_and_drop.rs]: Same.

File: src/drag_and_drop.rs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        let data = "I'm data!";
+        s.set_text(data, data.len() as i32);
+    });
+    button.connect_drag_data_delete(|w, d| {
+        println!("Source Drag Data Delete");
```
Comment (GuillaumeGomez) [inline src/drag_and_drop.rs]: Same.

File: src/drag_and_drop.rs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    // Configure label as drag destination
+    label.drag_dest_set(gtk::DEST_DEFAULT_ALL,
+                        &[],
+                        gdk::ACTION_COPY); // Maybe this should be ACTION_ASK?
+    label.connect_drag_motion(|w, d, _, _, _| {
+        println!("motion");
+        true
+    });
+    label.connect_drag_drop(|w, d, _, _, _| {
+        println!("drop");
```
Comment (GuillaumeGomez) [inline src/drag_and_drop.rs]: Same.

File: src/drag_and_drop.rs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        if !d.list_targets().is_empty() {
+            return true;
+        }
+        false
+    });
+    label.connect_drag_data_received(|w, d, _, _, s, _, _| {
+        println!("received");
```
Comment (GuillaumeGomez) [inline src/drag_and_drop.rs]: Same.

**Comment by Susurrus** [comment]
@GuillaumeGomez Indeed, this initial commit was primarily testing the library code I'm working on so you guys could see what's going on as I'm testing the changes to GTK and GDK.

Here's a smaller singular example. But when I run it I get a critical error about an Atom as soon as the GUI starts up. I'm not certain what exactly is going on here, but it looks like it's a failure in the Atom code I'm working on for GDK. @EPashkin any ideas?

```
(.:10245): Gdk-CRITICAL **: gdk_atom_intern: assertion 'atom_name != NULL' failed
```

**Comment by EPashkin** [comment]
@Susurrus it mean that you pass NULL pointer in this function.
Can you give me line that give you this error?

**Comment by Susurrus** [comment]
That's the entire error message I see on stdout, is there a way to get more detailed debugging information?

**Comment by Susurrus** [comment]
With the most recent changes I made to gtk-rs/gtk#472, I no longer get that error and I see the highlighting on the drop target, but `connect_drag_data_received` is never called, so I'm investigating that now.

**Comment by EPashkin** [comment]
You can add some `println("{}: {}", file!(), line!());` to detect lines that causes error

**Comment by Susurrus** [comment]
The most recent problem now is that setting the selection data doesn't work, calling `s.set_text` returns failure. I think this is why the `connect_drag_data_received` signal isn't being called, but I'm not certain about that. The code looks like:
```
    button.connect_drag_data_get(|_, _, mut s, _, _| {
        let data = "I'm data!";
        println!("{}", s.set_text(data, data.len() as i32));
    });
```

It'd also be nice is `s.set_text()` didn't need the length argument, but that's just a sidenote.

**Comment by GuillaumeGomez** [comment]
> It'd also be nice is `s.set_text()` didn't need the length argument, but that's just a sidenote.

That's actually an excellent point!

**Comment by EPashkin** [comment]
What returning `targets_include_text` before `set_text` attempt?
[Help](https://developer.gnome.org/gtk3/stable/gtk3-Selections.html#gtk-selection-data-set-text) says that text converted to target.
So need check `gtk_selection_data_get_target` and `gtk_selection_data_get_targets` too.

**Comment by Susurrus** [comment]
I'm not too familiar with how drag and drop works with GTK+ as their docs aren't very rich in this area, but I don't think that should be necessary because a) that isn't done in the example I'm porting over to Rust and b) these two are configured with identical text targets, so there shouldn't be any checking necessary.

**Comment by Susurrus** [comment]
For reference, this is the drag and drop example from the gtkmm project that I'm working off of: https://developer.gnome.org/gtkmm-tutorial/stable/sec-dnd-example.html.en

**Comment by Susurrus** [comment]
Looks like there might be an error with `gdk::DragContext::list_targets()` as it crashes when called in the latest iteration. The `drag-begin` and `drag-end` signals otherwise seem to be working properly though.

**Comment by EPashkin** [comment]
@Susurrus Maybe `list_target` now fixed with last glib.

**Comment by Susurrus** [comment]
@EPashkin `list_targets` is still crashing for me even with the latest glib.

**Comment by Susurrus** [comment]
Besides `list_targets` not working, `SelectionData::set_text()` isn't working within `WidgetExt::connect_drag_data_get()`. This is all that's required for this example, though the `list_targets` issue is still concerning.

**Comment by EPashkin** [comment]
@Susurrus, Main problem with both of your examples that it must based and pulled to `pending` branch.
Still looking why `list_targets` panics and why no drag-icon and result.

**Comment by EPashkin** [comment]
- `list_targets` can fixed with
```
impl GlibPtrDefault for Atom {
    type GlibType = ffi::GdkAtom;
}
```
but after this need checked all function accepts or returns arrays of `Atom`.
Currently don't remember any checked.
- `set_text` don't work due `targets_include_text` returning false
this IMHO can be partially fixed with
```
let list = button.drag_source_get_target_list().unwrap();
list.add_text_targets(0);
```
but really won't work. It added to list 7 text targets
```
UTF8_STRING
COMPOUND_TEXT
TEXT
STRING
text/plain;charset=utf-8
text/plain;charset=CP1251
text/plain
```
- main problem: `drag_source_set` add 2 `TargetEntry` with empty string. I still don't understand why.

**Comment by EPashkin** [comment]
Also if pass `&[]` to `drag_source_set` and `add_text_targets` after it
closure of `connect_drag_data_get` can change text (it checked by `get_text`) while `targets_include_text` still return false

**Comment by Susurrus** [comment]
I'll try to replicate what you've posted, as of right now I've pushed a rebased version so this is easier to work with. Not certain if you want me to open a new PR for merging this into the `pending` branch or not, so please let me know about that.

**Comment by EPashkin** [comment]
@Susurrus, Yes you need add PRs to `pending` branch` and close old.

**Pull request closed without merging**